packages/lesswrong/components/comments/CommentsTableOfContents.tsx

```diff
@@ -0,0 +1,189 @@
+
+const styles = (theme: ThemeType): JssStyles => ({
+  root: {
+    color: theme.palette.text.dim,
+  },
+  comment: {
+    display: "inline-flex",
+  },
+  commentKarma: {
+    width: 20,
+  },
+  commentAuthor: {
+  },
+  collapseButtonWrapper: {
+    marginLeft: 4,
+    height: 24,
+  },
+  collapseButton: {
+    "&:hover": {
+      background: theme.palette.greyAlpha(0.1),
+      borderRadius: 8,
+    },
+  },
+  collapseButtonExpanded: {
+  },
+  collapseButtonCollapsed: {
+    transform: "rotate(-90deg)",
+  },
+  postTitle: {
+    minHeight: 24,
+    paddingTop: 4,
+  },
+})
+
+const CommentsTableOfContents = ({commentTree, answersTree, post, classes}: {
+  commentTree?: CommentTreeNode<CommentsList>[],
+  answersTree?: CommentTreeNode<CommentsList>[],
+  post: PostsWithNavigation | PostsWithNavigationAndRevision,
+  classes: ClassesType,
+}) => {
+  const { TableOfContentsRow } = Components;
+  const flattenedComments = flattenCommentTree([
+    ...(answersTree ?? []),
+    ...(commentTree ?? [])
+  ]);
+  const { landmarkName: highlightedLandmarkName } = useScrollHighlight(
+    flattenedComments.map(comment => commentIdToLandmark(comment._id))
+  );
+  
+  if (!commentsTableOfContentsEnabled) {
+    return null;
+  }
+  
+  return <div className={classes.root}>
+    <TableOfContentsRow key="postTitle"
+      href="#"
+      onClick={ev => {
+        window.scrollTo({
+          top: 0,
+          behavior: "smooth"
+        });
+      }}
+      highlighted={highlightedLandmarkName==="above"}
+      title
+    >
+      <span className={classes.postTitle}>
+        {post.title?.trim()}
+      </span>
+    </TableOfContentsRow>
+
+    {answersTree && answersTree.map(answer => <>
+      <ToCCommentBlock
+        key={answer.item._id}
+        commentTree={answer} indentLevel={1} classes={classes}
+        highlightedCommentId={highlightedLandmarkName}
+      />
+      <Components.TableOfContentsDivider/>
+    </>)}
+    {commentTree && commentTree.map(comment => comment.item
+      ? <ToCCommentBlock
+          key={comment.item._id}
+          commentTree={comment} indentLevel={1} classes={classes}
+          highlightedCommentId={highlightedLandmarkName}
+        />
+      : null
+    )}
+  </div>
+}
+
+export function commentIdToLandmark(commentId: string): ScrollHighlightLandmark {
+  return {
+    landmarkName: commentId,
+    elementId: commentId,
+    position: "topOfElement",
+    offset: 25, //approximate distance from top-border of a comment to the center of the metadata line
+  }
+}
+
+const ToCCommentBlock = ({commentTree, indentLevel, highlightedCommentId, classes}: {
+  commentTree: CommentTreeNode<CommentsList>,
+  indentLevel: number,
+  highlightedCommentId: string|null,
+  classes: ClassesType,
+}) => {
+  const { TableOfContentsRow } = Components;
+  const { history } = useNavigation();
+  const location = useLocation();
+  const { query } = location;
+  const comment = commentTree.item!;
```

Comment:
`!`
😬 do we ever have a missing item in a comment node in practice?  (Is this for the case where we have comment nodes which are actually just visual separators, rather than comments, for e.g. comment permalinks?)